### PR TITLE
unit tests for miscellaneous functions in ovs.go

### PR DESCRIPTION
**Pull request by vjayaramrh**
Signed-off-by: Vishwanath Jayaraman <vjayaram@redhat.com>

**- What this PR does and why is it needed**

This PR adds unit tests for functions in pkg/util/ovs.go that depends on file access. The functions requiring file access are refactored to use the library https://github.com/spf13/afero that also make it easier for testing. The overview and features of afero libarary can be reviewed at https://github.com/spf13/afero/blob/master/README.md. The afero.NewMemMapFs() is used in testing and mocking file reads / writes.
The doc link for the APIs are at https://godoc.org/github.com/spf13/afero#Fs

This PR also removes the ginkgo test file pkg/util/ovs_test.go

Also, added a helper struct under pkg/testing to make it easy to structure tests involving file operations using afero

**- Special notes for reviewers**

**- How to verify it**

**- Description for the changelog**

### CODE REVIEW

File: go-controller/pkg/util/ovs_test.go [hunk 1]
```diff
@@ -1,180 +0,0 @@
...
-			Expect(err).NotTo(HaveOccurred())
-
-			app.Action = func(ctx *cli.Context) error {
-				err = SetExec(fexec)
-				Expect(err).NotTo(HaveOccurred())
-
-				_, err = config.InitConfig(ctx, fexec, nil)
-				Expect(err).NotTo(HaveOccurred())
-
-				args, env := getNbctlArgsAndEnv(10, "foo", "bar")
```
Comment (astoycos) [inline go-controller/pkg/util/ovs_test.go:78]: Is `getNbctlArgsAndEnv` the only function the existing `ovs_test` covered?

**Comment by trozet** [comment]
I'm not familiar with the afero library, but it looks decent. @danwinship or @dcbw do you have experience with it? What do you think?

**Comment by JacobTanenbaum** [comment]
is there any known performance issues with the afero library?

**Comment by vjayaramrh** [comment]
> Overall tests written using the go testing framework away from `ginkgo` is much more readable for a less experienced user , and it seems the test coverage in this case is much better. I.e the old test file `ovs_test.go` only seemed to cover one function in `ovs.go` titled `GetNbctlArgsandEnv` while the new unit tests cover the 9 shown below
> 
> ```
> RunningPlatform 
> TestRunOVNretry
> TestGetNbctlSocketPath
> GetNbctlArgsandEnv
> GetNbOVSDBArgs
> RunOVNNorthAppCtl
> RunOVNControllerAppCtl
> TestRunOvsVswitchdAppCtl
> DefaultExecRunner_RunCmd 
> ```
> 
> The most confusing part for a new user In my opinion would be the use of Mocks. The concept of Mocks is easy enough to understand, but in this instance the organization is somewhat confusing. For example, why are there mocks in `testing/mocks` and `util/mocks`? Also how they are called in the `ovs_unit_test.go` file gets a little confusing. Providing a short readme for the ovn-k8's repo describing how testing is implemented and organizationally laid out could be very helpful.

@astoycos appreciate you sharing your thoughts regarding mocks. In my earlier PR, I had shared my approach at https://github.com/ovn-org/ovn-kubernetes/pull/1371#discussion_r434186468 . Your suggestion about having a write up about mocks are organized is a good idea. I have opened an issue to track and address your comment at https://github.com/ovn-org/ovn-kubernetes/issues/1619

File: go-controller/pkg/util/ovs_unit_test.go [hunk 2]
```diff
@@ -25,6 +30,648 @@ type onCallReturnArgs struct {
...
+func TestRunOVNretry(t *testing.T) {
+	mockKexecIface := new(mock_k8s_io_utils_exec.Interface)
+	mockExecRunner := new(mocks.ExecRunner)
+	mockCmd := new(mock_k8s_io_utils_exec.Cmd)
+	// Below is defined in ovs.go
+	ovnCmdRetryCount = 0
+	// below is defined in ovs.go
+	runCmdExecRunner = mockExecRunner
+	// note runner is defined in ovs.go file
+	runner = &execHelper{exec: mockKexecIface}
```
[the remark this answers is not in the document]
Comment (vjayaramrh) [reply go-controller/pkg/util/ovs_unit_test.go:105]: They are not the same, refer https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/util/ovs.go#L170 and https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/util/ovs.go#L132

File: go-controller/pkg/util/ovs_test.go (hunk 1 above)
Comment (vjayaramrh) [reply go-controller/pkg/util/ovs_test.go:78]: your observation is correct

**Comment by vjayaramrh** [comment]
> is there any known performance issues with the afero library?

Going through their readme at https://github.com/spf13/afero it seems its more of an abstraction and layer(similiar to how we have interfaces or make calls to other libraries), does not look like there is a reason to be concerned about any performance issues.

**Comment by danwinship** [comment]
Yeah, I don't know anything about afero but it seems sane

**Pull request merged**